### ISSUE TITLE
Unable to implement classic COM interface on XAML component with C++/WinRT

### ISSUE THREAD

**Issue by sylveon** [microsoft/microsoft-ui-xaml#3331]
**Describe the bug**
When creating a XAML control/page, it's not possible to make it implement a classic COM interface (as C++/WinRT does it, by appending it to the base template arguments: `Something : SomethingT<Something, IMyClassicInterface>`) because the XAML compiler codegen does not account for this case.

**Steps to reproduce the bug**
1. Create a new blank C++/WinRT UWP app in VS
2. Open MainPage.h, and add the following code:
   - Include `<windows.ui.xaml.media.dxinterop.h>`.
   - Change the struct declaration to `struct MainPage : MainPageT<MainPage, ISwapChainPanelNative>`.
   - Add `IFACEMETHOD(SetSwapChain)(IDXGISwapChain*) noexcept;` to the struct members.
3. In MainPage.cpp, add the implementation:
   ```cpp
   IFACEMETHODIMP MainPage::SetSwapChain(IDXGISwapChain*) noexcept
   {
       // dummy implementation
       return S_OK;
   }
   ```
4. Try building the app.
5. Observe compiler errors.

**Expected behavior**
The app builds fine and a consumer is able to cast the XAML component to `ISwapChainPanelNative` and call the method provided by that interface.

**Screenshots**
![image](https://user-images.githubusercontent.com/6440374/94283802-78901900-ff1f-11ea-86fb-4efbe0e96e7e.png)

**Version Info**

NuGet package version: 
[Microsoft.Windows.CppWinRT 2.0.200921.6]

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| Yes | Yes |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2020 Update (19041)            | Yes   |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**
I replied Yes to the app type being Win32 because this issue showed up in system XAML islands while implementing `IInitializeWithWindow` to pass on a window handle to MessageDialog from my hosting code to the XAML code. I'm using MessageDialog because of ContentDialog's limitation of one dialog open at a time per thread. Implementing threading proved to be a huge pain, so I've decided on using MessageDialog for now.

The bug is caused by this line in MainPage.xaml.g.hpp:
```cpp
template struct MainPageT<struct MainPage>;
```

It tries to specialize `MainPageT` using different parameters than what I used in my declaration, and since the class uses CRTP, it tries to cast `this` back to MainPage, but because MainPage actually inherits from a different type (different template instantiations are effectively different types), the cast fails.

**Comment by StephenLPeters** [comment]
@kennykerr or @jevansaks I feel like we do this already, I think one of you two probably knows the answer. If not this sounds like a xaml compiler bug? @fabiant3

**Comment by StephenLPeters** [comment]
going with markup team for now.

**Comment by sylveon** [comment]
Yeah, it is a XAML compiler bug. Without XAML (a normal WinRT runtime component) it does work, but the XAML compiler's explicit specialization in .xaml.g.hpp is what breaks it.

**Comment by fabiant3** [comment]
@RealTommyKlein - can you take a look?

**Comment by sylveon** [comment]
My workaround was implementing the functionality in a subclass that doesn't use the XAML compiler: https://github.com/TranslucentTB/TranslucentTB/commit/8f68f3c0b1b1900b76c3900c98b8270da1e0c535

**Comment by DHowett** [comment]
Guessing that there's no workaround for this one? Just bit Terminal :/

**Comment by sylveon** [comment]
There's 2 options:
- putting this very ugly macro at the bottom of `.h` and undefining it in the `.cpp`
  `#define MyPage MyPage, IThing`
- Writing the code depending on a classic COM interface in a WinRT class not using the XAML Compiler (`Page` -> `PageWithThing` -> `MyPage`)

**Comment by kennykerr** [comment]
@Scottj1s any chance someone from the Xaml compiler can take a look at this seemingly trivial bug?

**Comment by DHowett** [comment]
@sylveon There was a brief but real moment when I realized I was here _because I was also implementing `IInitializeWithWindow`_ in a XAML islands application that I was quite crestfallen because _these are the depths to which we must sink._

Why does the XAML generated header need to forcibly instantiate this template anyway? Does it not suffice to let the consumer do it?

**Comment by sylveon** [comment]
The XAML codegen does this because some members are forward declared in the `.g.h` file, which you include, and then implemented + instantiated in `.g.hpp`, which you don't include (it's built as part of `XamlTypeInfo.g.cpp`).

**Comment by sylveon** [comment]
Without this explicit instantiation, you get linker errors.

**Comment by amtopel** [comment]
> Without this explicit instantiation, you get linker errors.

So does that mean this issue can never be fixed?

**Comment by sylveon** [comment]
Bump, still an issue

**Comment by amtopel** [comment]
> There's 2 options:
> 
> * putting this very ugly macro at the bottom of `.h` and undefining it in the `.cpp`
>   `#define MyPage MyPage, IThing`
> * Writing the code implementing the classic COM interface in a WinRT class not using the XAML Compiler (`Page` -> `PageWithThing` -> `MyPage`)

Could you clarify how you do the first workaround with the macros?  I've tried to do it but without success.  Can you show an example?  Thanks a lot.

**Comment by sylveon** [comment]
I really don't recommend it, it's extremely fragile.

**Comment by amtopel** [comment]
Another workaround might be to inherit from the classic COM interface outside of the `implements` template, disambiguate the queryinterface methods with `using`, and then handle a query for the interface in `query_interface_tearoff`.  I'm experimenting with that now.  It's still pretty ugly, because you have to cast away const-ness due to the `query_interface_tearoff` signature.

**Comment by amtopel** [comment]
It appears that modifying the declaration to this will compile, but it'll fail at runtime when `make` is called:
```
struct MyClass : implements<MyClass, MyClassT<MyClass>, ::IInitializeWithWindow>
{
   // ...
}
```

**Comment by amtopel** [comment]
@sylveon Maybe if you propose a specific way for them to fix this, there'll be a better chance of this getting fixed.  I'd propose one myself, but I don't understand the generated code (or C++ in general) nearly as well as you do.

**Comment by amtopel** [comment]
So far this is the cleanest workaround I've been able to come up with, but please feel free to suggest improvements.  You create only one XAML-based WinRT class.  In the implementation, you create a derived class (using `implements`) that implements any COM interfaces.  And then you specialize `make` to ensure that any outside code instantiating the class will actually instantiate the derived class.  That way any outside code doesn't need to change or know what's going on.

It looks like this:

```
namespace winrt::MyProject::implementation
{
    struct MyClass : MyClassT<MyClass>
    {
        MyClass();
    };

    struct MyClass_: implements<MyClass_, MyClass, ::ISomeCOMInterface>
    {
        // COM implementation
    };
}

namespace winrt
{
    template <>
    inline auto make<MyProject::implementation::MyClass>()
    {
        return make<MyProject::implementation::MyClass_>();
    }
}
```

**Comment by amtopel** [comment]
Why is this now categorized as a "feature proposal" instead of a bug?

**Comment by amtopel** [comment]
This bug also prevents you from adding any additional arguments whatsoever to the base class template--not just COM interfaces, but also marker structs, etc.

**Comment by amtopel** [comment]
I came up with a better workaround.  Let's say you have XAML class `MyClass`, and you want to implement `::IInitializeWithWindow`, for example.  What you do is insert a single-argument template specialization for `MyClassT` right before your class declaration, like so:

```
namespace winrt::MyProject::implementation
{
    template<typename T>
    struct MyClassT<T> : MyClassT<T, ::IInitializeWithWindow>
    {
    };
    
    struct MyClass : MyClassT<MyClass>
    {
        IFACEMETHOD(Initialize)(HWND) override;
    };
}
```

**Comment by amtopel** [comment]
This bug actually prevents the addition of any template arguments--not just COM interfaces.  For example, you can't add markers like `non_agile`, `composable`, etc.  Perhaps the title should be changed accordingly, and then this can be recategorized as a bug instead of a feature proposal.

**Comment by roxk** [comment]
Just hit this bug when adding supporting for `winrt::author::internal` in idlgen 2.0: https://github.com/roxk/idlgen/issues/114. Bump.
